<!DOCTYPE html>
<html>
<head>
<title>2024-05-04-MakeTreesInSlidingWindows.md</title>
<meta http-equiv="Content-type" content="text/html;charset=UTF-8">

<style>
/* https://github.com/microsoft/vscode/blob/master/extensions/markdown-language-features/media/markdown.css */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

body {
	font-family: var(--vscode-markdown-font-family, -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif);
	font-size: var(--vscode-markdown-font-size, 14px);
	padding: 0 26px;
	line-height: var(--vscode-markdown-line-height, 22px);
	word-wrap: break-word;
}

#code-csp-warning {
	position: fixed;
	top: 0;
	right: 0;
	color: white;
	margin: 16px;
	text-align: center;
	font-size: 12px;
	font-family: sans-serif;
	background-color:#444444;
	cursor: pointer;
	padding: 6px;
	box-shadow: 1px 1px 1px rgba(0,0,0,.25);
}

#code-csp-warning:hover {
	text-decoration: none;
	background-color:#007acc;
	box-shadow: 2px 2px 2px rgba(0,0,0,.25);
}

body.scrollBeyondLastLine {
	margin-bottom: calc(100vh - 22px);
}

body.showEditorSelection .code-line {
	position: relative;
}

body.showEditorSelection .code-active-line:before,
body.showEditorSelection .code-line:hover:before {
	content: "";
	display: block;
	position: absolute;
	top: 0;
	left: -12px;
	height: 100%;
}

body.showEditorSelection li.code-active-line:before,
body.showEditorSelection li.code-line:hover:before {
	left: -30px;
}

.vscode-light.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(0, 0, 0, 0.15);
}

.vscode-light.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(0, 0, 0, 0.40);
}

.vscode-light.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-dark.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 255, 255, 0.4);
}

.vscode-dark.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 255, 255, 0.60);
}

.vscode-dark.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-high-contrast.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 160, 0, 0.7);
}

.vscode-high-contrast.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 160, 0, 1);
}

.vscode-high-contrast.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

img {
	max-width: 100%;
	max-height: 100%;
}

a {
	text-decoration: none;
}

a:hover {
	text-decoration: underline;
}

a:focus,
input:focus,
select:focus,
textarea:focus {
	outline: 1px solid -webkit-focus-ring-color;
	outline-offset: -1px;
}

hr {
	border: 0;
	height: 2px;
	border-bottom: 2px solid;
}

h1 {
	padding-bottom: 0.3em;
	line-height: 1.2;
	border-bottom-width: 1px;
	border-bottom-style: solid;
}

h1, h2, h3 {
	font-weight: normal;
}

table {
	border-collapse: collapse;
}

table > thead > tr > th {
	text-align: left;
	border-bottom: 1px solid;
}

table > thead > tr > th,
table > thead > tr > td,
table > tbody > tr > th,
table > tbody > tr > td {
	padding: 5px 10px;
}

table > tbody > tr + tr > td {
	border-top: 1px solid;
}

blockquote {
	margin: 0 7px 0 5px;
	padding: 0 16px 0 10px;
	border-left-width: 5px;
	border-left-style: solid;
}

code {
	font-family: Menlo, Monaco, Consolas, "Droid Sans Mono", "Courier New", monospace, "Droid Sans Fallback";
	font-size: 1em;
	line-height: 1.357em;
}

body.wordWrap pre {
	white-space: pre-wrap;
}

pre:not(.hljs),
pre.hljs code > div {
	padding: 16px;
	border-radius: 3px;
	overflow: auto;
}

pre code {
	color: var(--vscode-editor-foreground);
	tab-size: 4;
}

/** Theming */

.vscode-light pre {
	background-color: rgba(220, 220, 220, 0.4);
}

.vscode-dark pre {
	background-color: rgba(10, 10, 10, 0.4);
}

.vscode-high-contrast pre {
	background-color: rgb(0, 0, 0);
}

.vscode-high-contrast h1 {
	border-color: rgb(0, 0, 0);
}

.vscode-light table > thead > tr > th {
	border-color: rgba(0, 0, 0, 0.69);
}

.vscode-dark table > thead > tr > th {
	border-color: rgba(255, 255, 255, 0.69);
}

.vscode-light h1,
.vscode-light hr,
.vscode-light table > tbody > tr + tr > td {
	border-color: rgba(0, 0, 0, 0.18);
}

.vscode-dark h1,
.vscode-dark hr,
.vscode-dark table > tbody > tr + tr > td {
	border-color: rgba(255, 255, 255, 0.18);
}

</style>

<style>
/* Tomorrow Theme */
/* http://jmblog.github.com/color-themes-for-google-code-highlightjs */
/* Original theme - https://github.com/chriskempson/tomorrow-theme */

/* Tomorrow Comment */
.hljs-comment,
.hljs-quote {
	color: #8e908c;
}

/* Tomorrow Red */
.hljs-variable,
.hljs-template-variable,
.hljs-tag,
.hljs-name,
.hljs-selector-id,
.hljs-selector-class,
.hljs-regexp,
.hljs-deletion {
	color: #c82829;
}

/* Tomorrow Orange */
.hljs-number,
.hljs-built_in,
.hljs-builtin-name,
.hljs-literal,
.hljs-type,
.hljs-params,
.hljs-meta,
.hljs-link {
	color: #f5871f;
}

/* Tomorrow Yellow */
.hljs-attribute {
	color: #eab700;
}

/* Tomorrow Green */
.hljs-string,
.hljs-symbol,
.hljs-bullet,
.hljs-addition {
	color: #718c00;
}

/* Tomorrow Blue */
.hljs-title,
.hljs-section {
	color: #4271ae;
}

/* Tomorrow Purple */
.hljs-keyword,
.hljs-selector-tag {
	color: #8959a8;
}

.hljs {
	display: block;
	overflow-x: auto;
	color: #4d4d4c;
	padding: 0.5em;
}

.hljs-emphasis {
	font-style: italic;
}

.hljs-strong {
	font-weight: bold;
}
</style>

<style>
/*
 * Markdown PDF CSS
 */

 body {
	font-family: -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif, "Meiryo";
	padding: 0 12px;
}

pre {
	background-color: #f8f8f8;
	border: 1px solid #cccccc;
	border-radius: 3px;
	overflow-x: auto;
	white-space: pre-wrap;
	overflow-wrap: break-word;
}

pre:not(.hljs) {
	padding: 23px;
	line-height: 19px;
}

blockquote {
	background: rgba(127, 127, 127, 0.1);
	border-color: rgba(0, 122, 204, 0.5);
}

.emoji {
	height: 1.4em;
}

code {
	font-size: 14px;
	line-height: 19px;
}

/* for inline code */
:not(pre):not(.hljs) > code {
	color: #C9AE75; /* Change the old color so it seems less like an error */
	font-size: inherit;
}

/* Page Break : use <div class="page"/> to insert page break
-------------------------------------------------------- */
.page {
	page-break-after: always;
}

</style>

<script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
</head>
<body>
  <script>
    mermaid.initialize({
      startOnLoad: true,
      theme: document.body.classList.contains('vscode-dark') || document.body.classList.contains('vscode-high-contrast')
          ? 'dark'
          : 'default'
    });
  </script>
<p>Navigate to /work/unckless/a948g501/SlidingTrees/ on the cluster</p>
<p>I'm using Devon's pipelines to make trees from my SNP vcf files.</p>
<p>I need an all sites vcf file rather than just SNPs so, I'm going to filter my raw vcf file for QC but use all called sites including invariants included</p>
<pre class="hljs"><code><div>bcftools filter -i <span class="hljs-string">'QUAL &gt; 30'</span> SNPs_ChrX_haploid.vcf -o filtered_allsites_ChrX_haploid.vcf

bcftools filter -i <span class="hljs-string">'QUAL &gt; 30'</span> SNPs_Autosomes_diploid.vcf -o filtered_allsites_Autosomes_diploid.vcf
</div></code></pre>
<p>The first step is to take the vcf files and convert it into a phylip file so that it can just be treated like any other DNA sequence alignment (although a rather large one).</p>
<p>Here's the code to do this -</p>
<pre class="hljs"><code><div><span class="hljs-comment">##get conversion script</span>
wget https://raw.githubusercontent.com/edgardomortiz/vcf2phylip/master/vcf2phylip.py

<span class="hljs-comment">##load python</span>
module load python

<span class="hljs-comment">##convert file</span>
python vcf2phylip.py -i /work/unckless/a948g501/SlidingTrees/filtered_allsites_ChrX_haploid.vcf
python vcf2phylip.py -i /work/unckless/a948g501/SlidingTrees/filtered_allsites_Autosomes_diploid.vcf
</div></code></pre>
<p>The PHYLIP matrix is saved to: filtered_allsites_ChrX_haploid.min4.phy and filtered_allsites_Autosomes_diploid.min4.phy</p>
<p>Now I will cut up the phylip</p>
<p>Credit to Jonothan Chang for a very clear explanation and framework for how best to do this (see: https://jonathanchang.org/blog/splitting-a-concatenated-raxml-style-phylip-file/)</p>
<p>For filtered_allsites_ChrX_haploid.min4.phy -</p>
<p>Length of the phylip alignment is 3,069,347</p>
<p>So I will generate 306 10Kb alignments and ignore the last 9.347Kb</p>
<p>To do so, I will define two variables in a for loop.</p>
<p>One (i) will track 1-306, over the 306 iterations<br>
one (j) will track 1-3,050,001 in increments of 10Kb, over the 306 iterations</p>
<p>The code '$((j+9999))' means add 9,999 to j each iteration, which will track 10,000-3,060,000 in increments of 10Kb, over the 306 iterations</p>
<p>Run this code to generate the necessary partition file</p>
<pre class="hljs"><code><div>j=<span class="hljs-number">1</span>
<span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> {<span class="hljs-number">1.</span><span class="hljs-number">.306</span>}
do
echo <span class="hljs-string">"DNA, p${i}=${j}-$((j+9999))"</span> &gt;&gt; partitions_X.txt
j=$((j+<span class="hljs-number">10000</span>))
done
</div></code></pre>
<p>For filtered_allsites_Autosomes_diploid.min4.phy -</p>
<p>Length of the phylip alignment is 6,033,053</p>
<p>So I will generate 603 10Kb alignments and ignore the last 3.053Kb</p>
<p>To do so, I will define two variables in a for loop.</p>
<p>One (i) will track 1-603, over the 603 iterations<br>
one (j) will track 1-6,020,001 in increments of 10Kb, over the 603 iterations</p>
<p>The code '$((j+9999))' means add 9,999 to j each iteration, which will track 10,000-6,030,000 in increments of 10Kb, over the 603 iterations</p>
<p>Run this code to generate the necessary partition file</p>
<pre class="hljs"><code><div>j=<span class="hljs-number">1</span>
<span class="hljs-keyword">for</span> i <span class="hljs-keyword">in</span> {<span class="hljs-number">1.</span><span class="hljs-number">.603</span>}
do
echo <span class="hljs-string">"DNA, p${i}=${j}-$((j+9999))"</span> &gt;&gt; partitions_Autosome.txt
j=$((j+<span class="hljs-number">10000</span>))
done
</div></code></pre>
<p>We will download the necessary python script to unconcatenate the phylip based on the above partition file</p>
<pre class="hljs"><code><div>curl -LO https://gist.githubusercontent.com/jonchang/<span class="hljs-number">34</span>c2e8e473ec2e8f50574671e62c3365/raw/unconcatenate_phylip.py
</div></code></pre>
<p>We will run the python script on this file to cut it up and read each subset file out into a new directory</p>
<pre class="hljs"><code><div>mkdir partitioned_X.phylips
module load python
python unconcatenate_phylip.py filtered_allsites_ChrX_haploid.min4.phy partitions_X.txt --prefix=partitioned_X.phylips/
</div></code></pre>
<pre class="hljs"><code><div>mkdir partitioned_Autosome.phylips
module load python
python unconcatenate_phylip.py filtered_allsites_Autosomes_diploid.min4.phy partitions_Autosome.txt --prefix=partitioned_Autosome.phylips/
</div></code></pre>
<p>The next step is just making a tree for each phylip alignment.</p>
<p>We are going to do this by submitting a 'batch job' on the cluster that spawns a unique job for each phylip alignment.</p>
<p>Here is my script for Script_RunGeneTreeArray_X.sh</p>
<pre class="hljs"><code><div><span class="hljs-comment">#!/bin/sh</span>
<span class="hljs-comment">#</span>
<span class="hljs-comment">#SBATCH --job-name=auto.gene.trees.X               </span>
<span class="hljs-comment">#SBATCH --nodes=1             </span>
<span class="hljs-comment">#SBATCH --ntasks-per-node=15               </span>
<span class="hljs-comment">#SBATCH --partition=sixhour            </span>
<span class="hljs-comment">#SBATCH --chdir=/work/unckless/a948g501/SlidingTrees/partitioned_X.phylips    </span>
<span class="hljs-comment">#SBATCH --mem-per-cpu=2gb            </span>
<span class="hljs-comment">#SBATCH --array=1-306</span>
<span class="hljs-comment">#SBATCH --time=360</span>


<span class="hljs-comment">#To be able to later use bootstrapping in ASTRAL, we will also generate sets of trees from bootstrapped gene alignments in the same IQ-TREE analysis, by using the option -B </span>
<span class="hljs-comment">#we do not specify a substition model with the -m option and therefore allow IQ-TREE to automatically select the best-fitting model.</span>
<span class="hljs-comment">#we will now ensure that the bootstrap trees are actually written to a file, and not just summarized in the output, by specifying the option --wbt.</span>
<span class="hljs-comment">#-nt AUTO ensures that all available CPU's are used</span>

<span class="hljs-comment">##load module iqtree</span>
module load iqtree
iqtree2 -s /work/unckless/a948g501/SlidingTrees/partitioned_X.phylips/DNA_p$SLURM_ARRAY_TASK_ID.phylip -bb <span class="hljs-number">1000</span> -wbt -nt AUTO
</div></code></pre>
<p>Make the Script executable -</p>
<pre class="hljs"><code><div>chmod +x Script_RunGeneTreeArray_X.sh
</div></code></pre>
<p>Here is my script for Script_RunGeneTreeArray_Autosome.sh</p>
<pre class="hljs"><code><div><span class="hljs-comment">#!/bin/sh</span>
<span class="hljs-comment">#</span>
<span class="hljs-comment">#SBATCH --job-name=auto.gene.trees.Autosome              </span>
<span class="hljs-comment">#SBATCH --nodes=1            </span>
<span class="hljs-comment">#SBATCH --ntasks-per-node=15              </span>
<span class="hljs-comment">#SBATCH --partition=sixhour            </span>
<span class="hljs-comment">#SBATCH --chdir=/work/unckless/a948g501/SlidingTrees/partitioned_Autosome.phylips    </span>
<span class="hljs-comment">#SBATCH --mem-per-cpu=2gb           </span>
<span class="hljs-comment">#SBATCH --array=1-603</span>
<span class="hljs-comment">#SBATCH --time=360</span>


<span class="hljs-comment">#To be able to later use bootstrapping in ASTRAL, we will also generate sets of trees from bootstrapped gene alignments in the same IQ-TREE analysis, by using the option -B </span>
<span class="hljs-comment">#we do not specify a substition model with the -m option and therefore allow IQ-TREE to automatically select the best-fitting model.</span>
<span class="hljs-comment">#we will now ensure that the bootstrap trees are actually written to a file, and not just summarized in the output, by specifying the option --wbt.</span>
<span class="hljs-comment">#-nt AUTO ensures that all available CPU's are used</span>

<span class="hljs-comment">##load module iqtree</span>
module load iqtree
iqtree2 -s /work/unckless/a948g501/SlidingTrees/partitioned_Autosome.phylips/DNA_p$SLURM_ARRAY_TASK_ID.phylip -bb <span class="hljs-number">1000</span> -wbt -nt AUTO
</div></code></pre>
<p>Make the Script executable -</p>
<pre class="hljs"><code><div>chmod +x Script_RunGeneTreeArray_Autosome.sh
</div></code></pre>
<p>This will create a lot of output files but the only thing I really need to worry about is the output tree file.<br>
Once I have all of the output trees, I use the cat command to paste them all into a single file.</p>
<pre class="hljs"><code><div>cat partitioned_X.phylips/DNA_p*.phylip.treefile &gt; OutputTrees_X.trees
</div></code></pre>
<pre class="hljs"><code><div>cat partitioned_Autosome.phylips/DNA_p*.phylip.treefile &gt; OutputTrees_Autosome.trees
</div></code></pre>
<p>Now we can calculate the proportion of trees supporting various topologies and visualize them using R</p>
<p>The R script for visualization can be found here - https://github.com/anjaligupta1210/AG_Unckless_Open_Lab_Notebook/blob/master/rscripts/SlidingTrees.R</p>

</body>
</html>
